internal/writeaheadlog/wal.go

```diff
@@ -0,0 +1,295 @@
+package writeaheadlog
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"slices"
+	"strings"
+	"sync"
+	"time"
+
+	logging "github.com/ipfs/go-log/v2"
+	cbg "github.com/whyrusleeping/cbor-gen"
+	"go.uber.org/multierr"
+)
+
+var log = logging.Logger("f3/wal")
+
+type Entry interface {
+	WALEpoch() uint64
+	cbg.CBORMarshaler
+	cbg.CBORUnmarshaler
+}
+
+type WriteAheadLog[T any, PT interface {
+	*T
+	Entry
+}] struct {
```

Comment:
I'm going put my rust police badge aside and ask: does this _need_ to be generic?

Reply:
I guess it doesn't really matter. I'm just wondering if it could be a bit simpler. If this makes testing simpler, that's fine.

Reply:
Yeah, makes testing simpler, I should also probably revert it and accept pointers.